File: leecode/401-500/475-供暖器.py
class Solution:
    def findRadius(self, houses, heaters) -> int:
        houses.sort()
        heaters.sort()
        m = len(houses)
        n = len(heaters)
        res = 0
        cur_heater = 0
        for house in houses:
            while cur_heater < n and heaters[cur_heater] < house:
                cur_heater += 1
            if cur_heater == 0:
                res = max(res, heaters[cur_heater] - house)
            elif cur_heater == n:
                return max(res, houses[-1] - heaters[-1])
            else:
                res = max(res, min(heaters[cur_heater]- house, house - heaters[cur_heater-1]))
        return res

File: leecode/401-500/475-供暖器.cpp
class Solution {
public:
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        sort(houses.begin(), houses.end());
        sort(heaters.begin(), heaters.end());
        int m = houses.size();
        int n = heaters.size();
        int res = 0;
        int cur_heater = 0;
        for (int& house : houses) {
            while (cur_heater < n && heaters[cur_heater] < house)
                cur_heater++;
            if (cur_heater == 0)
                res = max(res, heaters[cur_heater] - house);
            else if (cur_heater == n)
                return max(res, houses[m-1] - heaters[n-1]);
            else
                res = max(res, min(heaters[cur_heater]-house, house-heaters[cur_heater-1]));
        }
        return res;
    }
};